R/utils-get_code_dependency.R

```diff
@@ -0,0 +1,267 @@
+#' call. Influencers appear after `":"` string, e.g. `c("a", ":", "b", "c")` means a call influenced object named `a`,
+#' and the object was influenced by objects named `b` and `c`. If an object was marked with `@linksto` side effects tag
+#' then it appears as an influenced object at the beginning of a vector.
+#'
+#' @keywords internal
+#' @noRd
+code_graph <- function(calls_pd) {
+  cooccurence <- extract_occurrence(calls_pd)
+
+  side_effects <- extract_side_effects(calls_pd)
+
+  graph <- mapply(c, side_effects, cooccurence, SIMPLIFY = FALSE)
+
+  remove_graph_duplicates(graph)
+}
+
+#' Extract Object Occurrence
+#'
+#' @description Extract objects occurrence within calls passed by `calls_pd`. It also detects which objects are
+#' influenced by others within a call, and which are influencers.
+#'
+#' @param calls_pd A `list` of `data.frame`s, which is a result of `utils::getParseData()` grouped into separate calls.
+#' A result of `extract_calls()` function.
+#' @return A `list` (of length of input `calls_pd`) where each element represents one call. Each element consists of a
+#' character vector containing names of objects that were influenced by this call, and names of objects influencing this
+#' call. Influencers appear after `":"` string, e.g. `c("a", ":", "b", "c")` means a call influenced object named `a`,
+#' and the object was influenced by objects named `b` and `c`.
+#' @keywords internal
+#' @noRd
+extract_occurrence <- function(calls_pd) {
+  is_in_function <- function(x) {
+    # If an object is a function parameter,
+    # then in calls_pd there is a `SYMBOL_FORMALS` entry for that object.
+    function_id <- x[x$token == "FUNCTION", "parent"]
+    if (length(function_id)) {
+      x$id %in% get_children(x, function_id)$id
+    } else {
+      rep(FALSE, nrow(x))
+    }
+  }
+  lapply(
+    calls_pd,
+    function(call_pd) {
+
+      # What occurs in a function body is not tracked.
+      x <- call_pd[!is_in_function(call_pd), ]
+      sym_cond <- which(x$token %in% c("SYMBOL", "SYMBOL_FUNCTION_CALL"))
+
+      # Watch out for SYMBOLS after $ and @. For x$a x@a: x is object, a is not.
+      # For x$a, a's ID is $'s ID-2 so we need to remove all IDs that have ID = $ID - 2.
+      dollar_ids <- x[x$token %in% c("'$'", "'@'"), "id"]
+      if (length(dollar_ids)) {
+        object_ids <- x[sym_cond, "id"]
+        after_dollar <- object_ids[(object_ids - 2) %in% dollar_ids]
+        sym_cond <- setdiff(sym_cond, which(x$id %in% after_dollar))
+      }
+
+      # If there was an assignment operation detect direction of it.
+      ass_cond <- grep("ASSIGN", x$token)
+      if (length(ass_cond)) { # NOTE(1)
+        sym_cond <- sym_cond[sym_cond > ass_cond] # NOTE(2)
+      }
+      if ((length(ass_cond) && x$text[ass_cond] == "->") || !length(ass_cond)) { # NOTE(3)
+          sym_cond <- rev(sym_cond)
+      }
+      append(x[sym_cond, "text"], ":", after = 1)
```

Comment:
Just because it's short and contains a symbol not allowed in object names, I personally think that `:` is easier to be found in a vector of actual names, like the one below:

```r
graph_expected_4 <- list(
  c("iris2", ":", "iris"),
  c("iris3", "iris_head", ":", "head", "iris"),
  c("iris2", "iris3", ":", "iris_head"),
  c("classes", ":", "lapply", "iris2", "class")
)
```

when compares to

```r
graph_expected_4 <- list(
  c("iris2", "affected by", "iris"),
  c("iris3", "iris_head", "affected by", "head", "iris"),
  c("iris2", "iris3", "affected by", "iris_head"),
  c("classes", "affected by", "lapply", "iris2", "class")
)
```

Reply:
91327a9bc1592a4d58b104d5f71d4d9b15909121